src/main/groovy/edu/oregonstate/mist/locations/frontend/resources/LocationResource.groovy

```diff
@@ -82,21 +84,109 @@ class LocationResource extends Resource {
+        ObjectMapper mapper = new ObjectMapper() // can reuse, share globally
+        JsonNode actualObj = mapper.readTree(result)
+
+        def topLevelHits = actualObj.get("hits")
+        topLevelHits.get("hits").asList().each {
+            def source = it.get("_source")
+            ResourceObject resourceObject = new ResourceObject(
+                    id: source.get("id").asText(),
+                    type: source.get("type").asText(),
+                    attributes: source.get("attributes")
+            )
+            resultObject.data += resourceObject
+        }
+
+        setPaginationLinks(topLevelHits, q, type, campus, resultObject)
+
+        ok(resultObject).build()
+    }
+
+    /**
+     * Add pagination links to the data search results.
+     *
+     * @param topLevelHits          First "hits" node in the json document
+     * @param q
+     * @param type
+     * @param campus
+     * @param resultObject
+     */
+    private void setPaginationLinks(JsonNode topLevelHits, String q, String type, String campus,
+                                    ResultObject resultObject) {
+        def totalHits = topLevelHits.get("total").asInt()
+        // If no results were found, no need to add links
+        if (!totalHits) {
+            return
+        }
+
+        Integer pageNumber = getPageNumber()
+        Integer pageSize = getPageSize()
+        def map = ["q": q, "type": type, "campus": campus, "pageSize": pageSize, "pageNumber": pageNumber]
+        int lastPage = Math.ceil(totalHits / pageSize)
+
+        resultObject.links["self"] = getPaginationUrl(map)
+        map.pageNumber = 1
+        resultObject.links["first"] = getPaginationUrl(map)
+        map.pageNumber = lastPage
+        resultObject.links["last"] = getPaginationUrl(map)
+
+        if (pageNumber > DEFAULT_PAGE_NUMBER) {
+            map.pageNumber = pageNumber - 1
+            resultObject.links["prev"] = getPaginationUrl(map)
+        } else {
+            resultObject.links["prev"] = null
+        }
+
+        if (totalHits > (pageNumber * pageSize)) {
+            map.pageNumber = pageNumber + 1
+            resultObject.links["next"] = getPaginationUrl(map)
+        } else {
+            resultObject.links["next"] = null
+        }
+    }
+
+    /**
+     * Returns string url to use in pagination links.
+     *
+     * @param params
+     * @return
+     */
+    private String getPaginationUrl(def params) {
+        def uriAndPath = uriInfo.getBaseUri().toString() + uriInfo.getPath()
+        def nonNullParams = params.clone()
+        // convert pageVariable to page[variable]
+        nonNullParams["page[number]"] = nonNullParams['pageNumber']
+        nonNullParams["page[size]"] = nonNullParams['pageSize']
+        nonNullParams.remove('pageSize')
+        nonNullParams.remove('pageNumber')
+
+        // remove empty GET parameters
+        nonNullParams = nonNullParams.findAll { it.value } .collect { k, v -> "$k=$v" }
```

Comment:
I'm going to leave this as is. Here's why:
* Groovy actually expands findAll, but it uses metaprogramming and handles nulls
* The current performance of this API is 7ms , which is super fast. 
* I'm torn on converting this groovy code into loops because it adds more code to maintain, and if we end up using katharsis , we won't be needing this code.